### DISCUSSION TITLE
How to test nuxt composable useAsyncData in vitest

### DISCUSSION THREAD

**Discussion by SujithGunasekaran** [vitest-dev/vitest#2205] [Q&A]
Hi Team,

Recently I upgraded my nuxt 2 to nuxt 3. Currently, I am using this _3.0.0-rc.8_ version. In one of my .vue files, I use _useAsyncData_ hook to make an API call and store it in the pinia store please refer to the sample code it will look similar to the original code.  

```
<script setup>
  const store = useCartStore();
  
  // with help of useAsyncData.
  const { data, pending, error, refresh } = await useAsyncData(
    'getUserCart',
    () => store.getUserCart()
  );

  // I am using like this and it works in development. 
  await store.getUserCart();

useHead({});
definePageMeta({});

</script>

<template>
  <div v-if=store.userCard.length > 0>Code goes here...</div>
</template>
```

Now when I write a unit tests for this file using vitest. 

```
it('renders userCart Page', async () => {
    wrapper = shallowMount(UserCart, {
      global: {
        plugins: [
          i18n,
          router,
          store,
          createTestingPinia({
            createSpy: vi.fn,
            initialState: {
             userCart: {
               cart:  []
              }
            }
          })
        ],
        stubs,
      },
      props,
    });    
    expect(wrapper.vm).toBeTruthy();

    // below two is getting failed and when I console.log _wrappern.vm.definePageMeta_ getting undefined.
    expect(wrapper.vm.definePageMeta).toHaveBeenCalledOnce();
    expect(wrapper.vm.useHead).toHaveBeenCalledOnce();
  });
```

I ran into a problem and the component is not mounting since the setup function returns a promise. 
**_Component <Anonymous>: setup function returned a promise, but no <Suspense> boundary was found in the parent component tree. A component with async setup() must be nested in a <Suspense> in order to be rendered._**

can someone please help to resolve this issue? and Please tell me how we can write unit tests for this scenario.

**Comment by sheremet-va** [comment]
You need to wrap your component in Suspense: https://github.com/vuejs/test-utils/issues/108

**Reply by SujithGunasekaran** [reply]
Hi @sheremet-va, Thanks for your response. It works but I am using vue 3 composition API. When I try to access the state methods inside composition API it returns undefined.

```
<script setup>
const state = reactive({
 data: ''
})
</script>
```
Below is my test code 

```
it('renders UserCart Page', async () => {
    const mountSuspense = async (component, options) => {
      const wrapper = shallowMount(
        defineComponent({
          render() {
            return h(Suspense, null, {
              default: h(component),
              fallback: h('div', 'fallback')
            })
          }
        }), {
          ...options
        })
      await flushPromises()
      return wrapper;
    }
    wrapper = await mountSuspense(UserCart, {
      global: {
        plugins: [
          i18n,
          router,
          store,
          createTestingPinia({
            createSpy: vi.fn,
          })
        ],
        stubs,
      },
      props,
    });    
    
    console.log('hello', wrapper.vm); // gives undefined
});
```

**Reply by sheremet-va** [reply]
I don't know if it exists for script setup. Why would you access implementation details? It is generally considered a bad practice. I would suggest using testing library: https://testing-library.com/docs/guiding-principles/

**Reply by SujithGunasekaran** [reply]
Hi, @sheremet-va Thank you for your reply, I am not trying to access the implementation details, Actually, I am trying to access the DOM node but can't able to access, it because the component is not mounted.

**Reply by stikoo** [reply]
Hi @SujithGunasekaran, I'm also having trouble in this space, this little snippet over on the vitest docs has helped a little...

https://test-utils.vuejs.org/guide/advanced/async-suspense.html#testing-asynchronous-setup

I have something like this...
```
import LatestNews from '../LatestNews.vue'
import { createFullWrapper } from '~/tests/testUtilsHelper'
import { defineComponent, Suspense } from 'vue'

const TestComponent = defineComponent({
  components: { Suspense, LatestNews },
  props: {
    blok: Object
  },
  template: '<Suspense><LatestNews :blok="blok"/></Suspense>'
})

const props = {
  blok: {
    _uid: '12345',
    title: 'Latest News',
    errorMessage: `The thing is broken`,
    selectedNews: ['1', '2', '3']
  }
}

describe('LatestNews.test.js', () => {
  let testComponent
  let wrapper

  vi.mock('#imports', () => {
    return {
      default: { defaultKey: vi.fn() },
      useAsyncData(arg1, callback) {
        callback()
        return {
          data: 'data',
          pending: 'pending',
          error: 'error'
        }
      }
    }
  })


  it('Renders correctly', () => {
    wrapper = createFullWrapper(TestComponent, {
      props
    })

    expect(wrapper.exists()).toBe(true)
  })
})

```

The createFullWrapper helper function is just a wrapper around the test utils mount() method.

However, I'm still having trouble properly mocking the useAsyncData Nuxt composable inside `LatestNews.vue`, at the moment I've just got the above returning a basic object, I need to improve on this though to properly test the component.
